### PR TITLE
Mouse handler refactor

### PR DESCRIPTION
**Pull request by VehpuS**
Following https://github.com/maplibre/maplibre-gl-js/pull/1852#discussion_r1032895205, I've made this pull request to align the MouseHandler class to the factory style changes we made to the OneTouchHandler class. I also fixed a small bug in the instantiation of the OneTouchHandler class discovered during the refactor.

## Launch Checklist

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Manually test the debug page.

### CODE REVIEW

**Comment by VehpuS** [comment]
I have no clue why the integration tests are failing, and would appreciate any guidance on how to fix them...

**Comment by HarelM** [comment]
There's a problem with the tests, I'm not sure why and when it started.
I've seen some PRs that are failing for no good reason as well...
Might be a node version update if I need to guess... :-/

**Comment by VehpuS** [comment]
> 

That makes more sense than any of my changes having an effect :P 

If that's the case, I guess it's safer to wait for a solution for the tests before merging, right? Is the code otherwise acceptable?

**Comment by HarelM** [comment]
I think the code looks good, I need to fully look into it still, but before that I need to understand why the tests fail :-/

**Comment by HarelM** [comment]
Please rebase

File: src/ui/handler/mouse.ts [hunk 1]
```diff
@@ -45,12 +66,12 @@ class MouseHandler {
         delete this._eventButton;
     }
 
-    _correctButton(e: MouseEvent, button: number) {  //eslint-disable-line
-        return false; // implemented by child
-    }
-
-    _move(lastPoint: Point, point: Point): MouseMovementResult {  //eslint-disable-line
-        return {}; // implemented by child
+    _move(...params: Parameters<MouseMoveFunction>) {
```
[the remark this answers is not in the document]
Comment (VehpuS) [reply src/ui/handler/mouse.ts:69]: No, since doing so broke a lot of flows which assumed nothing is returned for no movement. But that can still be defined. I'll add it.

**Comment by VehpuS** [comment]
> Do we have tests that cover this code? If so, I would expect some changes there, if not, I think it would be a good idea to add some.

There are tests but they run on the map itself, not on the handler independently. Since I fixed the instantiation in the map's default handlers, the tests didn't need changing.

**Comment by HarelM** [comment]
I would consider adding unit tests around these class/classes to make sure the desired behavior is kept in the future.
Testing the handler by creating a test that uses the map is great but might not cover all line (100 - in the column just before the last means all lines of code are covered in tests).
For the mouse.ts file we see that it does when looking at the code coverage results:
image
But we see that for the others, it doesn't :-)

**Comment by VehpuS** [comment]
> I would consider adding unit tests around these class/classes to make sure the desired behavior is kept in the future. Testing the handler by creating a test that uses the map is great but might not cover all line (100 - in the column just before the last means all lines of code are covered in tests). For the mouse.ts file we see that it does when looking at the code coverage results: image But we see that for the others, it doesn't :-)

Do you want this done in this pull request?

**Comment by HarelM** [comment]
Overall, this looks great.
Some nit-picking:
I'm assuming the "_" fields are private, so I would consider removing them from the interface, if possible.
The two main interfaces turned out to be very similar (not surprisingly so), it might be interesting to use a single interface somehow, receiving the results and interaction (touch/move) event as part of the generic declaration.
Call the methods `interactionStart`, `interactionMoveWindow`, `...End`.
This is just a thought, I'm not sure how mature this thought is though...
Other than that, I think this looks very good.

**Comment by VehpuS** [comment]
Regarding the _ - it was usef externally before my refactor, I was considering removing the _ suffix as a fix but felt it was possibly breaking the existing interface to change this. I'll make the change then and hope the tests / type checks will catch any issues. Should I perhaps add a note in the changelog?

Regarding the common interface- this will change the API if both classes, though it does make sense. I can also create a class with the common logic (I would call it "dragging" interactions) and use it to implement the separate classes to avoid this interface change, though that might add more code than it saves. Do you have a preferred approach?

**Comment by HarelM** [comment]
These are very internal classes, I don't see a problem breaking their API. but sure, a note in the change log is needed.
I would avoid writing extra code in order to try keeping current behavior.
Note that we are preparing to release version 3, so breaking changes are allowed.

File: src/ui/control/navigation_control.ts [hunk 2]
```diff
@@ -152,23 +152,27 @@ class MouseRotateWrapper {
...
         this._clickTolerance = 10;
-        const mapRotateTolerance = map.dragRotate._mouseRotate._clickTolerance;
-        const mapPitchTolerance = map.dragRotate._mousePitch._clickTolerance;
+        const mapRotateTolerance = map.dragRotate._mouseRotate.getClickTolerance();
+        const mapPitchTolerance = map.dragRotate._mousePitch.getClickTolerance();
         this.element = element;
-        this.mouseRotate = new MouseRotateHandler({clickTolerance: mapRotateTolerance});
-        this.touchRotate = OneFingerTouchHandler.generateRotationHandler({clickTolerance: mapRotateTolerance});
+        this.mouseRotate = generateMouseRotationHandler({clickTolerance: mapRotateTolerance});
+        this.mouseRotate.enable();
```
Comment (HarelM) [inline src/ui/control/navigation_control.ts:167]: Consider adding this to the generate method if you always generate and then call `enable`.

File: src/ui/handler/drag_handler.ts [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        this._moveFunction = options.move;
+        this._activateOnStart = !!options.activateOnStart;
+
+        if (this._moveStateManager instanceof MouseMoveStateManager) {
+            if (options.preventContextMenu) {
+                this.contextmenu = function(e: E) {
+                    e.preventDefault();
+                };
+            }
+            this.mousedown = this.dragStart;
```
Comment (HarelM) [inline src/ui/handler/drag_handler.ts:75]: I don't understand this part, I'm probably missing something obvious - how is this assignment helpful?
Why does this class have extra properties - `touchstart`, `mousemove` etc, but they are not in use as part of the code in this class?

Comment (VehpuS) [reply src/ui/handler/drag_handler.ts:75]: The event handler calls them apparently (using the internal JS event names apparently) When they weren't implemented, I had 35 test failures, this solved all but the one that is failing. I couldn't find the code doing this either (otherwise I would likely be able to fix the last test too 😅)

Comment (HarelM) [reply src/ui/handler/drag_handler.ts:75]: I would search for strings with that name and binding methods. These are poorly used across the code and prevent "find all references" to work properly.
If you find those, please remove them and use the `dragStart` and like to bind to the relevant event.
This will also remove the need to check `instance of` in the constructor, which breaks the concept of interface and inheritance :-)

Comment (VehpuS) [reply src/ui/handler/drag_handler.ts:75]: True, though it was one step better than initializing with an enum 🙈. My problem was that I couldn't find the references in the code to dragstart/move/ebd which makes me concerned that this is reliant on browser APIs - but I will check that assumption further on my next round of work, based on your comments.

Comment (HarelM) [reply src/ui/handler/drag_handler.ts:75]: Nope, you need to register the event, somewhere, probably with `addEventListener(...)` or `on(...)`, no magic javascript :-)
As for enum, it still break the inheritance just the same. To avoid it you declare a method, `initialize` for example, and call that, each implementation will do things differently in that method.

Comment (VehpuS) [reply src/ui/handler/drag_handler.ts:75]: I wasn't suggesting an enum to be clear. I was saying my instanceof solution was meant to avoid using enum in the interface, but was clearly still a workaround to not finding the event listening logic 😅

I think I may have come across where the listeners are initialized, but if I recall this is done in a loop and assumes the existence of methods with the names of the events as a form of abstraction over the handlers. I'll check it the next coding session in any case.

Comment (HarelM) [reply src/ui/handler/drag_handler.ts:75]: Sounds like you found the right place. These assumptions are just bad because they cause runtime failures and not compile time failures, this is why I prefer to replace them as much as possible.

Comment (VehpuS) [reply src/ui/handler/drag_handler.ts:76]: `handler_manager.ts` is what calls the relevant handlers in the `handleEvent` method, and it is indeed using the broaser events as a source of the method names (and adding "window" events as well via the `handleWindowEvent` method, the source of the unit test failure). I don't see how to fundamentally change this without potentially affecting many other events... I will push a commit that improves the current implementation shortly, let me know if this is enough to consider this solved for this pull request. I'm happy try something more "extreme" if you think it's worth it, though IMO this might be better as a separate pull request.

**Comment by VehpuS** [comment]
FYI: I also ran the debug interface again, things seem to be running well as far as I can tell

File: src/ui/handler_manager.ts [hunk 4]
```diff
@@ -46,12 +46,15 @@ export interface Handler {
     // They are called with dom events whenever those dom evens are received.
     readonly touchstart?: (e: TouchEvent, points: Array<Point>, mapTouches: Array<Touch>) => HandlerResult | void;
     readonly touchmove?: (e: TouchEvent, points: Array<Point>, mapTouches: Array<Touch>) => HandlerResult | void;
+    readonly touchmoveWindow?: (e: TouchEvent, points: Array<Point>, mapTouches: Array<Touch>) => HandlerResult | void;
```
Comment (HarelM) [inline src/ui/handler_manager.ts:49]: Why are these needed? They are not DOM events...

Comment (VehpuS) [reply src/ui/handler_manager.ts:49]: They are generated by 'handleWindowEvent' in 'handler_manager.ts'

Comment (VehpuS) [reply src/ui/handler_manager.ts:49]: The fact that they were missing in the interface didn't cause an issue since, as I mentioned in another thread, they are called based on the string 'type' property of the browser events, and in the aforementioned method they are manipulated further using string interpolation...

File: src/ui/handler/drag_handler.ts (hunk 3 above)
Comment (HarelM) [reply src/ui/handler/drag_handler.ts]: Yea, I just understood what you wrote, this is just poor design and bad implementation that screams for bugs.
But it shouldn't be fixed as part of this PR.
This PR is good by itself. 
The handler registering code needs to change, avoid all these strings and function names - remove the need for "as any" which is showing there's a poor choice here etc. This reduction in code is not readable and doesn't help.
But, again, not part of this PR.

File: src/ui/handler_manager.ts (hunk 4 above)
Comment (HarelM) [reply src/ui/handler_manager.ts:49]: I see the problem now, the fact that they are all optional doesn't create a good interface though...
But again, not part of this PR...

**Review by HarelM — APPROVED** [review]

**Comment by VehpuS** [comment]
This issue can also be closed now: https://github.com/maplibre/maplibre-gl-js/issues/1933

**Comment by HarelM** [comment]
I've opened an issue to improve on handler manager: #1941
